### ISSUE TITLE
[DependencyInjection] Add attribute #[Exclude]

### ISSUE THREAD

**Issue by AlikDex** [symfony/symfony#46643]
### Description

Maybe we should add an Exclude attribute to exclude dto\vo\entity and other classes from registration in the service container?

Now I can do it with the When attribute. For example:
```
<?php

use Symfony\Component\DependencyInjection\Attribute\When;

#[When(env: 'never')]
class SomeClass
{
    // ...
}
```
But that seems like bad practice

### Example
```
<?php

use Symfony\Component\DependencyInjection\Attribute\Exclude;

#[Exclude]
class SomeClass
{
    // ...
}
```

**Comment by Eydamos** [comment]
Nice idea, but I don't like the name `Exclude` for that. Excluded from what? It is not very speaking and hard to follow for beginners. I would propose something like `NoService`

**Comment by nicolas-grekas** [comment]
> #[When(env: 'never')]
> But that seems like bad practice

Is it really? I quite like it: it both teaches/builds on existing knowledge about the `When` attr and reads nice to me.

**Comment by Eydamos** [comment]
> Is it really? I quite like it: it both teaches/builds on existing knowledge about the `When` attr and reads nice to me.

I think so too. Having something like `env:'never'` while the environment with the name `never` will actually never exist really feels like bad practice.
It is like: "Hey let me just put some random string here that might potentially give a hint that this will not exist just because I have no other way to excluded it from being a service"
Other people might be lazy and just do `env:"foobarbaz"` and a year later no one in the team knows any more why this was added.

**Comment by chalasr** [comment]
I agree it's too much of a hack :) 
Anyway, do we really want to put an attribute imported from the DependencyInjection component to all non-service classes? There's no plan to get rid of the default `services.yaml|php` file that enables PSR-4 based service discovery. Isn't adding a one-liner to that file simpler than putting that attribute to every single non-service class?

**Comment by ruudk** [comment]
> Isn't adding a one-liner to that file simpler than putting that attribute to every single non-service class?

That depends. It's often easy to exclude directories when you group files by type (form, controller, etc). But when you are not following that approach, it becomes very tedious to exclude certain files one by one. Especially in larger applications. I think it makes perfect sense to be able to exclude by attribute as we can almost do everything already via attributes.

**Comment by lyrixx** [comment]
Hello, actually #47196 did not fix this issue. So I'm re-opening it. There was 7 👍🏼 reaction on this issue, and I would like to see it in the core. I can submit a PR if you think it's OK.

> Anyway, do we really want to put an attribute imported from the DependencyInjection component to all non-service classes?

That's not an issue for me. I usually ignore all Exception, Model, Dto, etc folders. But for time to time I have static helpers that are not Symfony services. And I want to exclude them from the DIC.

**Comment by ruudk** [comment]
@lyrixx When you say "did not fix this issue" what do you mean? Because we are using this and it works great:

```php
<?php

declare(strict_types=1);

namespace App;

use Attribute;
use Symfony\Component\DependencyInjection\Attribute\When;

#[Attribute(Attribute::TARGET_CLASS | Attribute::TARGET_METHOD | Attribute::TARGET_FUNCTION | Attribute::IS_REPEATABLE)]
final class SkipAutowire extends When
{
    public function __construct()
    {
        parent::__construct('never');
    }
}
```

However, I would love to see this become a Framework attribute.

**Comment by lyrixx** [comment]
The title of this PR is : 
```
[DependencyInjection] Add attribute #[Exclude]
```
And there not `Exclude` attribute in the Framework (yet?). So, IMHO, the issue is not fixed.
